pimd/pim_sock.c

```diff
@@ -253,16 +311,65 @@ int pim_socket_mcast(int protocol, struct in_addr ifaddr, struct interface *ifp,
 	}
 
 	/* Set Tx socket DSCP byte */
-	if (setsockopt_ipv4_tos(fd, IPTOS_PREC_INTERNETCONTROL)) {
-		zlog_warn("can't set sockopt IP_TOS to PIM/IGMP socket %d: %s",
+	if (setsockopt_ipv6_tclass(fd, IPTOS_PREC_INTERNETCONTROL)) {
+		zlog_warn("can't set sockopt IP_TOS to PIMv6 socket %d: %s",
 			  fd, safe_strerror(errno));
 	}
 
+	return 0;
+}
+
+int pim_socket_mcast(int protocol, struct in_addr ifaddr, struct interface *ifp,
+		     uint8_t loop)
+{
+	int fd;
+	int ret;
+
+	fd = pim_socket_raw(protocol);
+	if (fd < 0) {
+		zlog_warn("Could not create multicast socket: errno=%d: %s",
+			  errno, safe_strerror(errno));
+		return PIM_SOCK_ERR_SOCKET;
+	}
+
+#ifdef SO_BINDTODEVICE
+
+	ret = pim_socket_bind(fd, ifp);
+	if (ret) {
+		close(fd);
+		zlog_warn(
+			"Could not set fd: %d for interface: %s to device",
+			fd, ifp->name);
+		return PIM_SOCK_ERR_BIND;
+	}
+#else
+/* XXX: use IP_PKTINFO / IP_RECVIF to emulate behaviour?  Or change to
+ * only use 1 socket for all interfaces? */
+#endif
+
+#if PIM_IPV == 4 
+	ret = pim_setsockopt_ipv4(protocol, fd, ifp, loop);
```

Comment:
rename `pim_setsockopt_ipv4` and `pim_setsockopt_ipv6` to `pim_setsockopt` so we don't need the `#if` here with both functions named the same

Reply:
Ack